File: src/word/Word.scala
package word

import scala.util.Random

/**
  * Created by k.neyman on 02.04.2017.
  */
abstract class Word(val word: String) {
  def this(word: List[Char]) = this(word.mkString)

  val size = word.length

  def apply(pos: Int): Char = {
    if (pos >= size) {
      println(word)
    }
    word(pos)
  }

  override def toString: String = word

  def next: Option[Word]
}

File: src/v_java/word/Word.java
package v_java.word;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Created by k.neyman on 03.04.2017.
 */
public abstract class Word {
    protected List<Character> word;
    protected int size;

    public Word(List<Character> list) {
        this.word = list;
        size = list.size();
    }

    public Character apply(int pos) {
        return word.get(pos);
    }

    @Override
    public String toString() {
        return word.stream().map(Object::toString).collect(Collectors.joining());
    }

    abstract Optional<Word> next();
}
